pkg/urbit/noun/events.c

```diff
@@ -638,18 +641,124 @@ _ce_image_sync(u3e_image* img_u)
   }
 }
 
-/* _ce_image_resize(): resize image, truncating if it shrunk.
- */
+//! Resize an image.
+//!
+//! Truncates an image if it shrunk.
+//!
+//! If the image shrunk and was mapped into memory, then the truncated portion
+//! is replaced with an anonymous, private (copy-on-write) mapping:
+//!
+//! ```
+//!  <high address>
+//! +==================+ <- previous end of image in memory
+//! |                  | <-+
+//! |                  |   |
+//! |------------------+   |
+//! |                  |   +- file-backed mapping replaced with anonymous
+//! |                  |   |  mapping
+//! |------------------    |
+//! |                  |   |
+//! |                  | <-+
+//! +==================+ <- new end of image in memory
+//!  <low address>
+//! ```
+//!
+//! If the image grew and was mapped into memory, then a new file-backed private
+//! (copy-on-write) mapping is established that is backed by the image:
+//!
+//! ```
+//!  <high address>
+//! +==================+ <- new end of image in memory
+//! |                  | <-+
+//! |                  |   |
+//! |------------------+   |
+//! |                  |   +- anonymous mapping replaced with file-backed
+//! |                  |   |  mapping
+//! |------------------|   |
+//! |                  |   |
+//! |                  | <-+
+//! +==================+ <- previous end of image in memory
+//! |                  |
+//! |                  |
+//! |      .....       |
+//! |                  |
+//! |                  |
+//! +==================+ <- base of image in memory
+//!  <low address>
+//! ```
+//!
+//! It's unclear whether anonymous mappings that lie within the previous bounds
+//! of the image (identified by `.....` in the diagram above) should be replaced
+//! with file-backed mappings when the image is resized. If the
+//! (Linux/macOS/Windows) kernel is smart enough to simply replace the
+//! anonymous mapping in the page cache with the file-backed mapping, then the
+//! anonymous mappings should be replaced with file-backed mappings. If, on the
+//! other hand, the kernel treats the pages of the two mappings as separate
+//! entities, then the cost of a cache miss may make replacing the anonymous
+//! mapping too expensive. For now, we remain conservative and don't replace the
+//! mappings and accept the potential cost of the accumulation of anonymous
+//! mappings (and the use of swap space that likely accompanies those mappings).
+//!
+//! @param[in] img_u  Image.
+//! @param[in] pgs_w  New size of the image.
+//! @param[in] bas_y  Base address of the image in memory. Used to establish a
+//!                   new mapping in memory. Should be NULL if no new mappings
+//!                   should be created.
+//!
+//! @n (1) The image is mapped into memory at base address `bas_y`.
+//! @n (2) The image shrunk.
+//! @n (3) The image grew.
```

Comment:
I've thought more about this style, and I still can't accept it. The nature of these comments is to granularly document specific parts of a function, but grouping them at the top and then referencing each inline forces the reader to manually dereference each comment. That indirection is cognitively costly, and the only reason for it is to make the comments legible to the doxygen machine. Machine-legibility is a good thing, but not worth damaging the readability or clarity of the code.

I object to those kind of tradeoffs in principle, but especially in this case: doxygen comments and generated artifacts are most appropriate for documenting interfaces for external consumption, but these comments are always about specific implementation details.

Reply:
I don't think dereferencing each comment is anymore cognitively expensive than some of the other tactics we use throughout the runtime, but at the same time I agree that the benefits don't outweigh the costs in this case. I've reverted back to a more conventional form of inline commenting.

Reply:
This is a good point, and I absolutely agree. There are some pretty cognitively-taxing patterns in those codebase: manual refcounting (and the style that makes it somewhat manageable), a very large surface of function-like macros (for concision or performance), and probably other things too.